downstream.go

```diff
@@ -2679,6 +2736,190 @@ func (dc *downstreamConn) handleMessageRegistered(ctx context.Context, msg *irc.
 				})
 			})
 		}
+	case "METADATA":
+		var target, subcommand string
+		if err := parseMessageParams(msg, &target, &subcommand); err != nil {
+			return err
+		}
+		switch subcommand {
+		case "SUB", "UNSUB":
+			for _, k := range msg.Params[2:] {
+				switch k {
+				case "pinned":
+				case "muted":
+				default:
+					dc.SendMessage(ctx, &irc.Message{
+						Command: "FAIL",
+						Params:  []string{msg.Command, "KEY_INVALID", k, "Invalid key"},
+					})
+					continue
+				}
+				switch subcommand {
+				case "SUB":
+					dc.metadataSubs[k] = struct{}{}
+					dc.SendMessage(ctx, &irc.Message{
+						Command: "770", // RPL_METADATASUBOK
+						Params:  []string{dc.nick, k},
+					})
+				case "UNSUB":
+					delete(dc.metadataSubs, k)
+					dc.SendMessage(ctx, &irc.Message{
+						Command: "771", // RPL_METADATAUNSUBOK
+						Params:  []string{dc.nick, k},
+					})
+				}
+			}
+			return nil
+		case "SUBS":
+			for k := range dc.metadataSubs {
+				dc.SendMessage(ctx, &irc.Message{
+					Command: "772", // RPL_METADATASUBS
+					Params:  []string{dc.nick, k},
+				})
+			}
+			return nil
+		}
+		if dc.network == nil || target == "*" {
+			return ircError{&irc.Message{
+				Command: "FAIL",
+				Params:  []string{"METADATA", "INVALID_TARGET", target, "Invalid target"},
+			}}
+		}
+		mt, err := dc.srv.db.GetMessageTarget(ctx, dc.network.ID, target)
+		if err != nil {
+			dc.logger.Printf("failed to get the message target for %q: %v", target, err)
+			return ircError{&irc.Message{
+				Command: "FAIL",
+				Params:  []string{"METADATA", "INTERNAL_ERROR", target, "Internal error"},
+			}}
+		}
+		switch subcommand {
+		case "LIST", "GET":
+			m := make(map[string]string)
+			if mt.Pinned {
+				m["pinned"] = "1"
+			} else {
+				m["pinned"] = "0"
+			}
+			if mt.Muted {
+				m["muted"] = "1"
+			} else {
+				m["muted"] = "0"
+			}
```

Comment:
Could extract this in a helper function since it's shared with `sendChannelMetadata`.